lib/nexthop_group.c

```diff
@@ -147,6 +148,42 @@ void copy_nexthops(struct nexthop **tnh, struct nexthop *nh,
 	}
 }
 
+uint32_t nexthop_group_hash(struct nexthop_group *nhg)
+{
+	struct nexthop *nh;
+	uint32_t i;
+	uint32_t key = 0;
+
+	/*
+	 * We are not interested in hashing over any recursively
+	 * resolved nexthops
+	 */
+	for (nh = nhg->nexthop; nh; nh = nh->next) {
+		key = jhash_1word(nh->type, key);
+		key = jhash_2words(nh->vrf_id, nh->nh_label_type, key);
+		/* gate and blackhole are together in a union */
+		key = jhash(&nh->gate, sizeof(nh->gate), key);
+		key = jhash(&nh->src, sizeof(nh->src), key);
+		key = jhash(&nh->rmap_src, sizeof(nh->rmap_src), key);
+		if (nh->nh_label) {
+			for (i = 0; i < nh->nh_label->num_labels; i++)
+				key = jhash_1word(nh->nh_label->label[i], key);
```

Comment:
I think it's worth special casing the 2 and 3word case, i.e., something like
i = nh->nh_label->num_labels;
while (i >= 3) {
   key = jhash_3words(nh->nh_label->label[i], nh->nh_label->label[i+1],nh->nh_label->label[i+2],key);
   i -=3;
}
if (i >= 2) {
   key = jhash_2words(nh->nh_label->label[i], nh->nh_label->label[i+1],key);
   i -=2;
}
if (i >= 2) {
   key = jhash_1word(nh->nh_label->label[i], key);
   i -=2;
}